src/transactions/ClawbackOpFrame.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2021 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "transactions/ClawbackOpFrame.h"
+#include "ledger/LedgerTxn.h"
+#include "transactions/TransactionUtils.h"
+#include <Tracy.hpp>
+
+namespace stellar
+{
+
+ClawbackOpFrame::ClawbackOpFrame(Operation const& op, OperationResult& res,
+                                 TransactionFrame& parentTx)
+    : OperationFrame(op, res, parentTx)
+    , mClawback(mOperation.body.clawbackOp())
+    , mAsset(getAsset())
+{
+}
+
+bool
+ClawbackOpFrame::isVersionSupported(uint32_t protocolVersion) const
+{
+    return protocolVersion >= 16;
+}
+
+bool
+ClawbackOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    ZoneNamedN(applyZone, "ClawbackOp apply", true);
+
+    LedgerKey key(TRUSTLINE);
+    key.trustLine().accountID = toAccountID(mClawback.from);
+    key.trustLine().asset = mAsset;
+
+    auto trust = ltx.load(key);
+    if (!trust)
+    {
+        innerResult().code(CLAWBACK_NO_TRUST);
+        return false;
+    }
+
+    if (!isClawbackEnabledOnTrustline(trust))
+    {
+        innerResult().code(CLAWBACK_NOT_CLAWBACK_ENABLED);
+        return false;
+    }
+
+    auto header = ltx.loadHeader();
+    if (!addBalanceSkipAuthorization(header, trust, -mClawback.amount))
+    {
+        innerResult().code(CLAWBACK_UNDERFUNDED);
+        return false;
+    }
+
+    innerResult().code(CLAWBACK_SUCCESS);
+    return true;
+}
+
+bool
+ClawbackOpFrame::doCheckValid(uint32_t ledgerVersion)
+{
+    if (mClawback.from == toMuxedAccount(getSourceID()))
+    {
+        innerResult().code(CLAWBACK_MALFORMED);
+        return false;
+    }
+
+    if (mClawback.amount < 1)
+    {
+        innerResult().code(CLAWBACK_MALFORMED);
+        return false;
+    }
+
+    if (!isAssetValid(mAsset))
+    {
+        innerResult().code(CLAWBACK_MALFORMED);
+        return false;
+    }
+
+    return true;
+}
+
+void
+ClawbackOpFrame::insertLedgerKeysToPrefetch(UnorderedSet<LedgerKey>& keys) const
+{
+    keys.emplace(trustlineKey(toAccountID(mClawback.from), mAsset));
+}
+
+Asset
+ClawbackOpFrame::getAsset() const
```

Comment:
A couple of comments about this:
- You might consider refactoring this into a non-member function in TransactionUtils because almost identical code appears in `AllowTrustOpFrame` twice
- You might consider making it throw in the `else` clause. Right now if we did implement another asset type and missed adding it here, it would result in a partially constructed asset which is definitely not good. Perhaps a `switch` statement would be better since it would warn about unimplemented branches.

Reply:
Moved and added an `else` clause that throws.